drivers/bt_ll_nrfxlib/entropy/entropy_ll_nrfxlib.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <entropy.h>
+#include <soc.h>
+#include <device.h>
+#include <irq.h>
+#include <kernel.h>
+#include <kernel_includes.h>
+#include <ble_controller.h>
+#include <ble_controller_soc.h>
+
+#include <multithreading_lock.h>
+
+/* The RNG driver data. */
+struct rng_driver_data {
+	/* Used to fill the pending client buffer with new random values after
+	 * RNG IRQ is triggered.
+	 */
+	struct k_sem sem_sync;
+};
+
+static struct rng_driver_data rng_data;
+
+static inline struct rng_driver_data *rng_driver_data_get(struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG((intptr_t) dev->driver_data == (intptr_t) &rng_data);
+
+	return (struct rng_driver_data *) (dev)->driver_data;
+}
+
+static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
+{
+	struct rng_driver_data *rng_dev = rng_driver_data_get(dev);
+	u8_t *p_dst = buf;
+	u32_t bytes_left = len;
+
+	while (bytes_left > 0) {
+		u32_t bytes_read = 0;
+
+		while (bytes_read == 0) {
+			int errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+			if (errcode) {
+				return errcode;
+			}
+			bytes_read = ble_controller_rand_vector_get(p_dst,
+								    bytes_left);
+			MULTITHREADING_LOCK_RELEASE();
+
+			if (!bytes_read) {
+				/* Put the thread on wait until next interrupt
+				 * to get more random values.
+				 */
+				k_sem_take(&rng_dev->sem_sync, K_FOREVER);
+			}
+		}
+
+		p_dst += bytes_read;
+		bytes_left -= bytes_read;
+	}
+
+	return 0;
+}
+
+static int rng_driver_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len,
+				      u32_t flags)
+{
+	int errcode = MULTITHREADING_LOCK_ACQUIRE_NO_WAIT();
```

Comment:
Idea behind the  entropy_get_entropy_isr() call is that there is separate entropy poll which could be accessed regardless of the state of main poll (accessible from the thread through separate API). So it always should return 0 or more bytes of data and no error (if the driver is initialized). To be sure that you always get the data, you can use ENTROPY_BUSYWAIT flag, which force the call to spin while waiting for the entropy.

Here the situation is tricky:
- ENTROPY_BUSYWAIT is not checked, so we always doing busy-wait even if the caller is not ready for the delay.
- Due to locking scheme this function will fail if anything else took the lock. Even if called with ENTROPY_BUSYWAIT flag...
- As this function is supposed to be called from ISR, caller cannot implement any kind of retry mechanism to handle failure and ensure that the random data could be obtained (this is classic priority inversion problem, but we cannot reschedule IRQ and thread in order to mitigate this).

Reply:
Also, please note that  entropy_get_entropy_isr() is re-entrant by the design, as IRQ could be pre-empted by a higher-priority IRQ...

Reply:
We've added some documentation explaining the difference in behavior:
https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/687/files#diff-3408bae23cb79d87a923a686e0dd82cf

It's not ideal and a solution for this needs to be worked out, but it will not be done for ble-controller v0.2.0.